src/hooks/useProducts.js

```diff
@@ -0,0 +1,61 @@
+import { db } from '../lib/firebase';
+import { collection, query, onSnapshot, addDoc } from '@firebase/firestore';
+import { useState, useEffect } from 'react';
+
+export const useProducts = () => {
+  const [products, setProducts] = useState([]);
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState(false);
+
+  const addProduct = async () => {
```

Comment:
It's interesting 😮!  You have an "extra" function in your custom hook: `addProduct`. 

This works! But, I would like to ask you: 

❓ Did you consider moving your `addProduct` function to **another file** like `helpers` or **isolate this logic** from your custom hook?

Reply:
Thanks! I understood all the points, but I still had a doubt. Why does the fetching data function need to be a custom hook? Couldn't it stay in `firebaseUtils` too? In other words, I wonder if it's a convention that the fetching data function is a custom hook while the other operations don't have to be?

Reply:
You're welcome! Good question as well. Thank you 💯 !

About your question: _"Why does the fetching data function need to be a custom hook?"_ I'm not sure, but I guess Yancarlo & you (@IsabellaSoares)  decided to create a custom hook 😄, and you believed this is the best way to solve this issue 🥳  .  it's great you have created it💯 !

I didn't find a rule or policy that it says if you need to fetch data, you should create a custom hook.  But I support custom hooks when I ask myself: 

- Do I need to use some results in multiple places? -> **Extract logic, and share it.**
- Do I need to have my own states? (Your current code: `loading` state is reused, but I understood is a state when your fetching data, you could lose some control)
- Is my component difficult to read, and maintain? (of course: vision about the future, but avoid over-engineering)
- Do I want to have more control over **my logic, and isolation goals**? (For example: when is `loading`, fetching data, and happened an `error`)

Also, 👀 if we pay attention to small details your custom hook could be divide into more parts: `firebaseOperations.js` (CRUD), `firebaseHelpers.js` (transform Firebase docs to JS objects, filters, remove elements, and so on), and your `useProducts.js` could be simpler, and other great things. Maybe, you can do in the next issues 🚀 😄.

**Bonus.** There are other external libraries that follow this **unofficial "pattern"** like React Router, Redux, and so on. I believe it has good acceptance in React community.

**Happy for your curiosity and interest!** 🥳 😄

Reply:
Wow! This is very interesting! 👀 We hadn't thought of all these possibilities to resolve the issue. It's the first time we've both implemented a custom hook, but the idea of ​​separating responsibilities into different files is also pretty cool! Thank you for the help! We will evaluate the possibilities and make the necessary corrections! 🚀